docs/proposals/20191016-e2e-test-framework.md

```diff
@@ -0,0 +1,199 @@
+* Provide test-infra / prow integration. This is left up to each provider as not all providers will be inside the kubernetes or kubernetes-sigs organization.
+
+## Proposal
+
+The crux of this proposal is to implement a test suite/framework/library that providers can use as e2e tests. It will not cover provider specific edge cases. Providers are still left to implement their own provider specific e2es.
+
+This framework will capture behaviors of Cluster API that span all providers.
+
+### User Stories
+
+#### As a Cluster API provider implementor I want to know my provider behaves in a way that Cluster API expects
+
+By providing a pluggable framework that contains generic Cluster API behavior for the provider implementor to use, we are able to satisfy this user story.
+
+#### As a developer in the Cluster API ecosystem I want to have confidence that my change does not break expected behaviors
+
+By writing down the expected behaviors of Cluster API in the form of a test framework, we will be able to know if our changes break those expected behaviors and thus satisfy this user story.
+
+#### As a Cluster API developer, I want to be sure I’m not accidentally disrupting existing providers.
+
+Existing providers will be depending on Cluster API behavior. This is an alternative perspective of the previous user story.
+
+#### As the release manager of a Cluster API provider, I want to be able to recommend new releases to users as soon as they come up
+
+Not all providers have automated testing that use all components of Cluster API and test interactions. This framework would be able to provide that signal to release managers to know if a release will at least pass this suite.
+
+#### As a user of Cluster API, I want to be sure new releases I’m installing won’t cause regressions.
+
+Regressions are easy to introduce. If we happen to introduce a regression that is not covered by these behaviors we will be able to identify how to write a test to reproduce the bad behavior and to fix the underlying bug.
+
+### Implementation Details/Notes/Constraints
+
+This framework will be implemented using the Ginkgo/Gomega behavioral testing framework because that is the Kubernetes standard and does a good job of structuring and organizing behavior based tests.
+
+This framework will live in the `sigs.k8s.io/cluster-api` module because it will be formalizing behaviors of Cluster API. Therefore the set of behaviors for each version may change, so will the test suite.
+
+This framework is not conformance nor will be built in the same style as the conformance framework of kubernetes. That is to say, there will not be a distributed binary. The only artifact needed is Cluster API's go module appropriately tagged and pushed.
+
+These tests will be consumed by a provider just like any other go library dependency. They can be imported and provider custom types can be passed in and the e2es will be run for the given provider, assuming secrets are also managed
```

Comment:
Are we hashing out how the custom types are passed into the e2e tests in this CAEP or something for the actual PR? Sounds like it will dramatically impact the design of the framework and something that should be considered early

Reply:
The interface is likely going to be `runtime.Object`. Since one of the goals is to be provider agnostic, if this framework does not work for any provider that we agree it should work for, then that goal is failed.

The exact set of functions will surface as we start implementing it. Does that work for you? Is there anything in particular I can help address?